api/agent/agent.go

```diff
@@ -807,6 +742,107 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
 	logger.WithError(err).Info("hot function terminated")
 }
 
+// runHotReq enqueues a free slot to slot queue manager and watches various timers and the consumer until
+// the slot is consumed. A return value of false means, the container should shutdown and no subsequent
+// calls should be made to this function.
+func (a *agent) runHotReq(ctx context.Context, call *call, state ContainerState, logger logrus.FieldLogger, cookie drivers.Cookie, slot *hotSlot) bool {
+
+	var err error
+	isFrozen := false
+
+	freezeTimer := time.NewTimer(a.freezeIdleMsecs)
+	idleTimer := time.NewTimer(time.Duration(call.IdleTimeout) * time.Second)
+	ejectTicker := time.NewTicker(a.ejectIdleMsecs)
+
+	defer freezeTimer.Stop()
+	defer idleTimer.Stop()
+	defer ejectTicker.Stop()
+
+	// log if any error is encountered
+	defer func() {
+		if err != nil {
+			logger.WithError(err).Error("hot function failure")
+		}
+	}()
+
+	// if an immediate freeze is requested, freeze first before enqueuing at all.
+	if a.freezeIdleMsecs == time.Duration(0) && !isFrozen {
+		err = cookie.Freeze(ctx)
+		if err != nil {
+			return false
+		}
+		isFrozen = true
+	}
+
+	state.UpdateState(ctx, ContainerStateIdle, call.slots)
+	s := call.slots.queueSlot(slot)
+
+	for {
+		select {
+		case <-s.trigger: // slot already consumed, unfreeze ASAP
+			if !isFrozen {
+				err = cookie.Freeze(ctx)
+				if err != nil {
+					return false
+				}
+				isFrozen = true
+			}
+		case <-ctx.Done(): // container shutdown
+		case <-a.shutdown: // server shutdown
+		case <-idleTimer.C:
+		case <-freezeTimer.C:
+			if !isFrozen {
+				err = cookie.Freeze(ctx)
+				if err != nil {
+					return false
+				}
+				isFrozen = true
+			}
+			continue
+		case <-ejectTicker.C:
+			// if someone is waiting for resource in our slot queue, we must not terminate,
+			// otherwise, see if other slot queues have resource waiters that are blocked.
+			stats := call.slots.getStats()
+			if stats.containerStates[ContainerStateWait] > 0 ||
+				a.resources.GetResourceTokenWaiterCount() <= 0 {
+				continue
+			}
+			logger.Debug("attempting hot function eject")
+		}
+		break
+	}
+
+	// if we can acquire token, that means we are here due to
+	// abort/shutdown/timeout, attempt to acquire and terminate,
+	// otherwise continue processing the request
+	if call.slots.acquireSlot(s) {
+		slot.Close(ctx)
+		return false
+	}
+
+	// In case, timer/acquireSlot failure landed us here, make
+	// sure to unfreeze.
+	if isFrozen {
+		err = cookie.Unfreeze(ctx)
+		if err != nil {
+			return false
+		}
+		isFrozen = false
+	}
+
+	state.UpdateState(ctx, ContainerStateBusy, call.slots)
+
+	// request processing may take a long time, clean up timers now
+	ejectTicker.Stop()
+	freezeTimer.Stop()
+	idleTimer.Stop()
+
+	// wait for this call to finish
+	// NOTE do NOT select with shutdown / other channels. slot handles this.
+	<-slot.done
```

Comment:
one option to do this (`<-slot.done`) in the caller, which would remove the double ticker/timer stops